File: src/org/cappuccino_project/ide/intellij/plugin/psi/ObjJElementFactory.java
package org.cappuccino_project.ide.intellij.plugin.psi;

import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiErrorElement;
import com.intellij.psi.PsiFileFactory;
import com.intellij.psi.util.PsiTreeUtil;
import org.cappuccino_project.ide.intellij.plugin.annotator.IgnoreUtil;
import org.cappuccino_project.ide.intellij.plugin.lang.ObjJFile;
import org.cappuccino_project.ide.intellij.plugin.lang.ObjJLanguage;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement;
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class ObjJElementFactory {
    private static final Logger LOGGER = Logger.getLogger(ObjJElementFactory.class.getName());
    public static final String PLACEHOLDER_CLASS_NAME = "_XXX__";

    @Nullable
    public static ObjJSelector createSelector(Project project, String selector) {
        if (selector == null || selector.isEmpty()) {
            return null;
        }
        String scriptText = "@implementation "+PLACEHOLDER_CLASS_NAME+" \n - (void) "+selector+"{} @end";
        ObjJImplementationDeclaration implementationDeclaration = (ObjJImplementationDeclaration)createFileFromText(project, scriptText).getClassDeclarations().get(0);
        if (implementationDeclaration.getMethodDeclarationList().isEmpty()) {
            return null;
        } else if (implementationDeclaration.getMethodDeclarationList().get(0).getMethodHeader().getSelectorList().isEmpty()) {
            return null;
        }
        return implementationDeclaration.getMethodDeclarationList().get(0).getMethodHeader().getSelectorList().get(0);
    }

    public static ObjJVariableName createVariableName(Project project, String variableName) {
        String scriptText = "var "+variableName+";";
        ObjJFile file = createFileFromText(project, scriptText);
        ObjJBodyVariableAssignment variableAssignment = ObjJTreeUtil.getChildOfType(file, ObjJBodyVariableAssignment.class);
        assert variableAssignment != null;
        return variableAssignment.getQualifiedReferenceList().get(0).getVariableNameList().get(0);
    }

    @Nullable
    public static ObjJFunctionName createFunctionName(@NotNull Project project, @NotNull String functionName) {
        String scriptText = String.format("function %s(){}", functionName);
        LOGGER.log(Level.INFO, "Script text: <"+scriptText+">");
        ObjJFile file = createFileFromText(project, scriptText);
        ObjJFunctionDeclaration functionDeclaration = ObjJTreeUtil.getChildOfType(file, ObjJFunctionDeclaration.class);
        return functionDeclaration != null ? functionDeclaration.getFunctionName(): null;
    }

    public static PsiElement createSpace(Project project) {
        String scriptText = " ";
        ObjJFile file = createFileFromText(project, scriptText);
        return file.getFirstChild();
    }

    public static PsiErrorElement createSemiColonErrorElement(Project project) {
        String scriptText = "?*__ERR_SEMICOLON__*?";
        ObjJFile file = createFileFromText(project, scriptText);
        ObjJErrorSequence errorSequence = ObjJTreeUtil.getChildOfType(file, ObjJErrorSequence.class);
        PsiErrorElement errorElement = ObjJTreeUtil.getChildOfType(errorSequence, PsiErrorElement.class);
        if (errorElement == null) {
            List<String> childElementTypes = new ArrayList<>();
            for (PsiElement child : file.getChildren()) {
                childElementTypes.add(child.getNode().getElementType().toString());
            }
            LOGGER.log(Level.INFO, "createSemiColonErrorElement(Project project) Failed. No error element found. Found <"+ ArrayUtils.join(childElementTypes)+"> instead");
        }
        return errorElement;
    }

    @Nullable
    public static ObjJComment createIgnoreComment(Project project, IgnoreUtil.ElementType elementType) {
        String scriptText = "//ignore "+elementType.type;
        ObjJFile file = createFileFromText(project, scriptText);
        return PsiTreeUtil.getChildOfType(file, ObjJComment.class);
    }

    private static ObjJFile createFileFromText(Project project, String text) {
        return (ObjJFile) PsiFileFactory.getInstance(project).createFileFromText("dummy.j", ObjJLanguage.INSTANCE, text);
    }

    @NotNull
    private static ObjJMethodDeclaration createMethodDeclaration(@NotNull Project project, @NotNull ObjJMethodHeader methodHeader) {
        String script = "@implementation "+PLACEHOLDER_CLASS_NAME+" \n " +
                methodHeader.getText() + "\n" +
                "{" + "\n" +
                "    " + getDefaultReturnValueString(methodHeader.getMethodHeaderReturnTypeElement()) + "\n" +
                "}" + "\n" +
                "@end";
        ObjJFile file = createFileFromText(project, script);
        ObjJImplementationDeclaration implementationDeclaration = file.getChildOfType(ObjJImplementationDeclaration.class);
        assert implementationDeclaration != null;
        ObjJMethodDeclaration declaration = implementationDeclaration.getMethodDeclarationList().get(0);
        assert declaration != null;
        return declaration;
    }

    private static String getDefaultReturnValueString(@Nullable ObjJMethodHeaderReturnTypeElement returnTypeElement) {
        if (returnTypeElement == null || returnTypeElement.getFormalVariableType() == null) {
            return "";
        }
        String defaultValue = "nil";
        final ObjJFormalVariableType formalVariableType = returnTypeElement.getFormalVariableType();
        if (formalVariableType.getVarTypeBool() != null) {
            defaultValue = "NO";
        } else if (formalVariableType.getVarTypeId() != null || formalVariableType.getClassName() != null) {
            defaultValue = "Nil";
        }
        return "return " + defaultValue + ";";
    }

    @NotNull
    public static ObjJReturnStatement createReturnStatement(@NotNull Project project, String returnValue) {
        String scriptString =
                "function x() {\n" +
                    "return " + returnValue + ";" + "\n" +
                "}";
        ObjJFile file = createFileFromText(project, scriptString);
        ObjJFunctionDeclarationElement functionDeclarationElement = file.getChildOfType(ObjJFunctionDeclarationElement.class);
        assert functionDeclarationElement != null;
        ObjJReturnStatement returnStatement = functionDeclarationElement.getChildOfType(ObjJReturnStatement.class);
        assert returnStatement != null;
        return returnStatement;
    }

    public static PsiElement createCRLF(Project project) {
        final ObjJFile file = createFileFromText(project, "\n");
        return file.getFirstChild();
    }

}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/ObjJElementFactory.kt
package org.cappuccino_project.ide.intellij.plugin.psi

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiErrorElement
import com.intellij.psi.PsiFileFactory
import com.intellij.psi.util.PsiTreeUtil
import org.cappuccino_project.ide.intellij.plugin.annotator.IgnoreUtil
import org.cappuccino_project.ide.intellij.plugin.lang.ObjJFile
import org.cappuccino_project.ide.intellij.plugin.lang.ObjJLanguage
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement
import org.cappuccino_project.ide.intellij.plugin.utils.ArrayUtils
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJTreeUtil

import java.util.ArrayList
import java.util.logging.Level
import java.util.logging.Logger

object ObjJElementFactory {
    private val LOGGER = Logger.getLogger(ObjJElementFactory::class.java.name)
    val PLACEHOLDER_CLASS_NAME = "_XXX__"

    fun createSelector(project: Project, selector: String?): ObjJSelector? {
        if (selector == null || selector.isEmpty()) {
            return null
        }
        val scriptText = "@implementation $PLACEHOLDER_CLASS_NAME \n - (void) $selector{} @end"
        val implementationDeclaration = createFileFromText(project, scriptText).classDeclarations[0] as ObjJImplementationDeclaration
        if (implementationDeclaration.methodDeclarationList.isEmpty()) {
            return null
        } else if (implementationDeclaration.methodDeclarationList[0].methodHeader.selectorList.isEmpty()) {
            return null
        }
        return implementationDeclaration.methodDeclarationList[0].methodHeader.selectorList[0]
    }

    fun createVariableName(project: Project, variableName: String): ObjJVariableName {
        val scriptText = "var $variableName;"
        val file = createFileFromText(project, scriptText)
        val variableAssignment = ObjJTreeUtil.getChildOfType(file, ObjJBodyVariableAssignment::class.java)!!
        return variableAssignment.qualifiedReferenceList[0].variableNameList[0]
    }

    fun createFunctionName(project: Project, functionName: String): ObjJFunctionName? {
        val scriptText = String.format("function %s(){}", functionName)
        LOGGER.log(Level.INFO, "Script text: <$scriptText>")
        val file = createFileFromText(project, scriptText)
        val functionDeclaration = ObjJTreeUtil.getChildOfType(file, ObjJFunctionDeclaration::class.java)
        return functionDeclaration?.functionName
    }

    fun createSpace(project: Project): PsiElement {
        val scriptText = " "
        val file = createFileFromText(project, scriptText)
        return file.firstChild
    }

    fun createSemiColonErrorElement(project: Project): PsiErrorElement? {
        val scriptText = "?*__ERR_SEMICOLON__*?"
        val file = createFileFromText(project, scriptText)
        val errorSequence = ObjJTreeUtil.getChildOfType(file, ObjJErrorSequence::class.java)
        val errorElement = ObjJTreeUtil.getChildOfType(errorSequence, PsiErrorElement::class.java)
        if (errorElement == null) {
            val childElementTypes = ArrayList<String>()
            for (child in file.children) {
                childElementTypes.add(child.node.elementType.toString())
            }
            LOGGER.log(Level.INFO, "createSemiColonErrorElement(Project project) Failed. No error element found. Found <" + ArrayUtils.join(childElementTypes) + "> instead")
        }
        return errorElement
    }

    fun createIgnoreComment(project: Project, elementType: IgnoreUtil.ElementType): ObjJComment? {
        val scriptText = "//ignore " + elementType.type
        val file = createFileFromText(project, scriptText)
        return PsiTreeUtil.getChildOfType(file, ObjJComment::class.java)
    }

    private fun createFileFromText(project: Project, text: String): ObjJFile {
        return PsiFileFactory.getInstance(project).createFileFromText("dummy.j", ObjJLanguage.INSTANCE, text) as ObjJFile
    }

    private fun createMethodDeclaration(project: Project, methodHeader: ObjJMethodHeader): ObjJMethodDeclaration {
        val script = "@implementation " + PLACEHOLDER_CLASS_NAME + " \n " +
                methodHeader.text + "\n" +
                "{" + "\n" +
                "    " + getDefaultReturnValueString(methodHeader.methodHeaderReturnTypeElement) + "\n" +
                "}" + "\n" +
                "@end"
        val file = createFileFromText(project, script)
        val implementationDeclaration = file.getChildOfType(ObjJImplementationDeclaration::class.java)!!
        return implementationDeclaration.methodDeclarationList[0]!!
    }

    private fun getDefaultReturnValueString(returnTypeElement: ObjJMethodHeaderReturnTypeElement?): String {
        if (returnTypeElement == null || returnTypeElement.formalVariableType == null) {
            return ""
        }
        var defaultValue = "nil"
        val formalVariableType = returnTypeElement.formalVariableType
        if (formalVariableType!!.varTypeBool != null) {
            defaultValue = "NO"
        } else if (formalVariableType.varTypeId != null || formalVariableType.className != null) {
            defaultValue = "Nil"
        }
        return "return $defaultValue;"
    }

    fun createReturnStatement(project: Project, returnValue: String): ObjJReturnStatement {
        val scriptString = "function x() {\n" +
                "return " + returnValue + ";" + "\n" +
                "}"
        val file = createFileFromText(project, scriptString)
        val functionDeclarationElement = file.getChildOfType(ObjJFunctionDeclarationElement<*>::class.java)!!
        return functionDeclarationElement.getChildOfType(ObjJReturnStatement::class.java)!!
    }

    fun createCRLF(project: Project): PsiElement {
        val file = createFileFromText(project, "\n")
        return file.firstChild
    }

}
